Fix useAngle logic to accurately render the angle

rickyrombo opened the pull request:
Fixes the useAngle calculations so that they accurately render the proper angle, matching the web implementation.

Ports the [Chromium implementation](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/css/css_gradient_value.cc;l=883-952;bpv=1;bpt=1?q=linear-gradient%20lang:c%2B%2B&ss=chromium%2Fchromium%2Fsrc) to Java and Obj-C. I am not a Obj-C dev (first time), feedback greatly appreciated.

Concerns:
- How will the Chromium license affect consumers of the package? Do I need to include the Chromium copyright notice?
- This will break existing usage as it completely changes the way the points are calculated. Probably warrants a breaking change/major version?

In-depth writeup on the maths here: https://understandable.dev/deep-dives/react-native-linear-gradient/

Addresses #416 and #425

rickyrombo commented:
cc: @danielgindi @dgladkov Tagging some people here to get some thoughts and eyes on this!

danielgindi commented:
I have one simple question: Why do we need to have an implementation identical to Chromium? And even copy their code?..

rickyrombo commented:
We don't have to copy Chromium. I can try my best to rewrite the math in a different way, but functionally I think it will have to mirror what Chromium did anyway in order to yield the expected result. The influence will be there no matter what I do. Changing variable names and moving functions around feels like copying someone's writing for a page in my novel but changing the character names and rearranging a few paragraphs. The math problem here is direct and I'm not creative enough to see another way to solve it elegantly. If all you're asking is for maybe using the gradient line equation instead of the perpendicular line for finding `endY`, and/or solving for the start point and reflecting to find the end point, and/or using 90 degrees = west and moving some variables around and renaming/reorganizing things, totally can do that.

It's possible that by porting and adapting the code it's already different enough that it no longer counts as the same "source code" and needs no copyright/attribution - but I am not a lawyer and I wanted to be forthcoming as possible about my sources. 😬

If your question is more aimed at why the implementation should match _web_ in general, I guess I'd say it doesn't have to (and with this implementation, it's still slightly different as we have `angleCenter`), but to me it makes a great deal of sense, as many devs are following designs from tools like Figma and/or trying to apply similar styles between mobile platforms and web, so it would be of great benefit if they could "learn once, write anywhere." Note that the web version of this package is inconsistent already: https://github.com/react-native-web-community/react-native-web-linear-gradient/blob/master/src/index.js

I am genuinely curious if I understand your current implementation correctly - perhaps I don't have the right mental model for the goal it's trying to achieve. Maybe you could help me understand? As it is currently, if you set the angle to 135 degrees, the gradient starts at (0, 0) and ends at (width, height). That's not what I would expect - it doesn't look like 135 degrees for a 4x2 rectangle, for example. That will only look like 135 degrees (at least, what I would expect 135 degrees to look like) if `width = height`, ie. in a square.

danielgindi commented:
The current implementation works kind of like photoshop.
Where you can set the center for the gradient, and an angle.
The angle is correctly used and the gradient is calculated dynamically based on the size of the layer and the angle+center selected.
This solves the original issue of the gradient being distorted by the width-height ratio as the default was to have a start/end points.
Now if you have a photoshop file and want to make a gradient exactly like in the psd- you could use the center and angle like in the psd.

If you want to mimick web behavior, you have so much more going on that you did not copy from chrome... Someone trying to copy gradient code from css would be frustrated with a web-compatible code that does not allow them more than 2% of what the web actually allows.

rickyrombo commented:
Sounds like we want the same thing - we want an easy way to convert from the design to the code. In my case, that's Figma, in yours, it seems it's Photoshop.

I didn't have Photoshop so I got the free trial to see if it matches. I made the following gradient in Photoshop:
![image](https://user-images.githubusercontent.com/3690498/153330650-fcc02f32-eba1-4897-b59a-df5abc5163f7.png)
![image](https://user-images.githubusercontent.com/3690498/153330694-3d128ed0-43be-4eb3-85cf-4da5aa5459af.png)
![image](https://user-images.githubusercontent.com/3690498/153330824-62edd34d-dfe7-47c2-9367-870ee4f47b4c.png)

Here's the equivalent `react-native-linear-gradient` code, as I would expect it:
```tsx
import React from 'react';
import {SafeAreaView, StyleSheet} from 'react-native';
import LinearGradient from 'react-native-linear-gradient';

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#a0c4ff',
  },
  gradient: {
    width: 320,
    height: 60,
  },
});

const App = () => {
  return (
    <SafeAreaView style={styles.container}>
      <LinearGradient
        useAngle
        angle={45}
        angleCenter={{x: 0.5, y: 0.5}}
        colors={['#FFF', '#000']}
        locations={[0.5, 0.5]}
        style={styles.gradient}
      />
    </SafeAreaView>
  );
};

export default App;
```
Yet it renders:
![image](https://user-images.githubusercontent.com/3690498/153331109-c3b6c692-9ce3-4b05-88b0-955a0415c6d5.png)

Which is different than what Photoshop produced. **This is what this PR fixes.** 

With this PR, the output looks like:
![image](https://user-images.githubusercontent.com/3690498/153331453-37b2cf12-5992-49b3-b79b-85dc23540a75.png)

This is what developers should expect.

> The angle is correctly used and the gradient is calculated dynamically based on the size of the layer and the angle+center selected.
> This solves the original issue of the gradient being distorted by the width-height ratio as the default was to have a start/end points.

Everything is calculated accurately until the final step of multiplying by the width and height of the original element, as that gets distributed to the angleSize as well as the angleCenter, causing the skew. I explain this in length on this blog post: https://understandable.dev/deep-dives/react-native-linear-gradient/ - have you taken a peek at it?

```java
        mShader = new LinearGradient(
                startPos[0] * mSize[0], // start X
                startPos[1] * mSize[1], // start Y
                endPos[0] * mSize[0], // end X
                endPos[1] * mSize[1], // end Y
            mColors,
            mLocations,
            Shader.TileMode.CLAMP);
```

danielgindi commented:
I agree that there is a bug :) I also agree of course that this should be fixed.
I'm just not convinced that adding the complexity of web's gradients- without it's actual features- is necessary and will actually add any feature.
In css you can use the gradients to create really weird shapes and patterns. The native gradients in both iOS and android do not support that.

Also if we do decide to go this way- we should supply some code to translate the old simple angle property into the new way (in a correct manner of course, without the skew).

Full disclosure:
I'm the one who implemented this back when I needed a native gradient (without calculating it in js and async rendering it, which caused a glitch).
The tests were done on really colorful gradients, with transitions not as sharp as this. And of course an iPhone's screen, not a long strip :) So I guess visually I missed the fact that there's a skew. 
I'm not using this feature anymore, but I believe that many do, and that support for the current feature should be maintained, and that a major version release is required to do it rendering differently after the fix.

rickyrombo commented:
I agree with the concern, we don't want to have the CSS-like string `linear-gradient(...)` for example - that would confuse developers who think they can put in the same exact gradient as they would on web and be frustrated with the results.

I don't do that here - I focused on maintaining the current contract for the end user and only updating the internal calculations so that the final output shown above comes from the same JS code - using the props `useAngle`, `angle` and `angleCenter`. I don't think this adds any additional complexity nor tries to bring all the web features to the package, but it does change the math to match the way the web (and photoshop) calculates the start and end points based on an angle. I probably muddied the water a bit by saying I ported the Chromium implementation - what I meant was I did the start/end point calculations the same way as Chromium. Does that address your concerns?

I agree that this would require a major version release, as it would be a breaking change for people that have been working around it. :100:

PS: Very, very much appreciate your initial contribution of the angle support! Thank you for helping solve the flickering issue. I agree that a lot of people are using it and it's definitely important to support. By the way, totally understand how this was missed - it's _very_ easy to miss this with even a slightly more squarish render area and an actual, colorful gradient rather than black and white with hard stops.

danielgindi commented:
Now we need to understand what the implications are on the license of this library.
And perhaps there's another way to  fix the issue? I mean it's just angles, some trigonometry...

rickyrombo commented:
Indeed! Something tells me all we need to do is calculate the math with minor differences to the code to remove the need for the notice entirely. There's no way they can copyright the math, right? Let me put up another commit this weekend with some more home-grown code.

rickyrombo commented:
I updated the PR, removing the copyright notices and changing the math and code to be more like the existing code structure. Tested on iOS and Android simulators.

Is this repo still active? Who has write access?

danielgindi commented:
Yes this looks much better :)

danielgindi commented:
@dulmandakh

friederbluemle commented on android/src/main/java/com/BV/LinearGradient/LinearGradientView.java:
```diff
@@ -97,41 +97,133 @@ protected void onSizeChanged(int w, int h, int oldw, int oldh) {
+     * @param size the size of the element
+     * @return the start point, relative to the angle center in cartesian
+     */
+    static private float[] getHorizontalOrVerticalStartPoint (float angle, int[] size) {
```
If these 3 new methods should be static, can you please move them further up in the file? Right below the constructor on line 36/37 would be good.
The simple reason is because that's the default sorting order in Android Studio, and we want to avoid a major formatting-only commit that changes all these lines.
Also: There should be no space after the method name and `(`.

friederbluemle commented on android/src/main/java/com/BV/LinearGradient/LinearGradientView.java:
```diff
@@ -97,41 +97,133 @@ protected void onSizeChanged(int w, int h, int oldw, int oldh) {
+     */
+    static private float[] getHorizontalOrVerticalStartPoint (float angle, int[] size) {
+        float halfWidth = size[0] / 2;
+        float halfHeight = size[1] / 2;
```
This right here triggers "integer division in floating-point context".

Can you change it to either `size[0] / 2f` or `size[0] / 2.0` to remove the warning? Same applies to the two identical lines in `getStartCornerToIntersect`.

friederbluemle reviewed (commented):
Hi @rickyrombo - Thanks for the PR! I'm happy to get this merged. I don't have detailed knowledge in the logic/math changes in this PR but it looks like @danielgindi has also reviewed the PR and signed off on it. Thank you!

I left a couple of minor comments, can you please check and address them?

Also: Right now there are 4 commit on the PR head branch - including one that adds a 7000 line package-lock.json file and another that removes it again. Can you please squash the fixup commits accordingly and rebase the branch? To me it looks like a single commit should be ok for this PR.

rickyrombo commented:
Thanks so much for the review @friederbluemle ! I ran Android Studio's "Reformat" to fix the formatting issues requested, fixed the division warning, and squashed the commits. Also rebased on latest master.

friederbluemle commented:
Thanks @rickyrombo :+1: Can you please rebase one more time? I've merge (most) of the formatting change in a separate PR (#549)
I purposefully excluded two individual hunks from `LinearGradientView.java` in order to avoid conflicts with your branch. Your commit should apply cleanly on top of the latest master branch.

rickyrombo commented:
@friederbluemle done! Thanks!

rickyrombo commented:
@friederbluemle let me know if there's anything more you need from me here! :)

friederbluemle reviewed (approved):
Looks great, thanks for the contribution @rickyrombo 👏
